File: bucket_2/1.py
# child_twitter.proto

from google.protobuf import field_mask_pb2

# Account
class Account:
    ACCOUNT_TYPE_UNSPECIFIED = 0
    PARENT_ACCOUNT = 1
    CHILD_ACCOUNT = 2

    ACCOUNT_LIFECYCLE_UNSPECIFIED = 0
    PENDING_EMAIL_VERIFICATION = 1
    PENDING_SIGN_UP = 2
    ACTIVE = 3
    DELETE_REQUESTED = 4

class MAccount:
    name: str
    type: Account.AccountType
    account_lifecycle: Account.AccountLifecycle
    birthdate: str
    email: str
    password: str

class MAddParentRequest:
    account: str
    parent_account: str
    second_parent_email: str

class MAddParentResponse:
    account: MAccount

# Comment
class MComment:
    name: str
    text: str

class MCreateAccountRequest:
    account: MAccount
    account_id: str

class MCreateCommentRequest:
    parent: str
    comment: MComment

class MCreateTweetRequest:
    parent: str
    tweet: MTweet

class MDeleteTweetRequest:
    name: str

class MExportTweetsRequest:
    account: str
    filename: str

class MExportTweetsResponse:
    pass  # No fields here

class MFollowAccountRequest:
    name: str

class MFollowAccountResponse:
    followed_accounts: list[str]

class MLikeTweetRequest:
    name: str

class MLikeTweetResponse:
    pass  # No fields here

class MListTweetsRequest:
    parent: str
    page_size: int
    page_token: str
    order_by: str

class MListTweetsResponse:
    tweets: list[MTweet]
    next_page_token: str

class MReTweetRequest:
    name: str

class MReTweetResponse:
    tweet: MTweet

class MSearchAccountsRequest:
    name: str

class MSearchAccountsResponse:
    accounts: list[str]

class MTweet:
    name: str
    text: str
    likes: int
    original_tweet: str

class MUpdateAccountRequest:
    account: MAccount
    update_mask: field_mask_pb2.FieldMask

# TwitterForKidsService
class STwitterForKidsService:
    def CreateAccount(self, request: MCreateAccountRequest) -> MAccount:
        pass

    def UpdateAccount(self, request: MUpdateAccountRequest) -> MAccount:
        pass

    def AddParent(self, request: MAddParentRequest) -> MAddParentResponse:
        pass

    def SearchAccounts(self, request: MSearchAccountsRequest) -> MSearchAccountsResponse:
        pass

    def FollowAccount(self, request: MFollowAccountRequest) -> MFollowAccountResponse:
        pass

    def CreateTweet(self, request: MCreateTweetRequest) -> MTweet:
        pass

    def DeleteTweet(self, request: MDeleteTweetRequest) -> google.protobuf.Empty:
        pass

    def ListTweets(self, request: MListTweetsRequest) -> MListTweetsResponse:
        pass

    def LikeTweet(self, request: MLikeTweetRequest) -> MLikeTweetResponse:
        pass

    def ReTweet(self, request: MReTweetRequest) -> MReTweetResponse:
        pass

    def CreateComment(self, request: MCreateCommentRequest) -> MComment:
        pass

    def ExportTweets(self, request: MExportTweetsRequest) -> MExportTweetsResponse:
        pass






#CLIENT SIDE:
import grpc
from child_twitter_pb2 import MCreateTweetRequest, MDeleteTweetRequest, MLikeTweetRequest, MCreateCommentRequest
from child_twitter_pb2_grpc import STwitterForKidsServiceStub

# Create a gRPC channel to connect to the server
channel = grpc.insecure_channel('localhost:50051')  # Replace with your server's address

# Create a stub for the TwitterForKidsService
twitter_service = STwitterForKidsServiceStub(channel)

# Create a tweet
def create_tweet(parent, tweet_text):
    tweet = MCreateTweetRequest(parent=parent, tweet=MTweet(text=tweet_text))
    response = twitter_service.CreateTweet(tweet)
    print(f"Tweet created: {response.tweet.name}")

# Delete a tweet
def delete_tweet(tweet_name):
    tweet = MDeleteTweetRequest(name=tweet_name)
    twitter_service.DeleteTweet(tweet)
    print(f"Tweet deleted: {tweet_name}")

# Like a tweet
def like_tweet(tweet_name):
    like_request = MLikeTweetRequest(name=tweet_name)
    twitter_service.LikeTweet(like_request)
    print(f"Tweet liked: {tweet_name}")

# Reply to a tweet
def reply_to_tweet(parent_tweet_name, reply_text):
    reply = MCreateCommentRequest(parent=parent_tweet_name, comment=MComment(text=reply_text))
    response = twitter_service.CreateComment(reply)
    print(f"Reply created: {response.name}")

# usage
if __name__ == "__main__":
    create_tweet("parent_account", "Hello, Twitter for Kids!")
    delete_tweet("tweet123")
    like_tweet("tweet456")
    reply_to_tweet("tweet789", "That's great!")

File: bucket_2/1.ts
// child_twitter.proto

enum AccountType {
    ACCOUNT_TYPE_UNSPECIFIED = 0,
    PARENT_ACCOUNT = 1,
    CHILD_ACCOUNT = 2,
  }
  
  enum AccountLifecycle {
    ACCOUNT_LIFECYCLE_UNSPECIFIED = 0,
    PENDING_EMAIL_VERIFICATION = 1,
    PENDING_SIGN_UP = 2,
    ACTIVE = 3,
    DELETE_REQUESTED = 4,
  }
  
  interface MAccount {
    name: string;
    type: AccountType;
    account_lifecycle: AccountLifecycle;
    birthdate: string;
    email: string;
    password: string;
  }
  
  interface MAddParentRequest {
    account: string;
    parent_account: string;
    second_parent_email: string;
  }
  
  interface MAddParentResponse {
    account: MAccount;
  }
  
  interface MComment {
    name: string;
    text: string;
  }
  
  interface MCreateAccountRequest {
    account: MAccount;
    account_id: string;
  }
  
  interface MCreateCommentRequest {
    parent: string;
    comment: MComment;
  }
  
  interface MCreateTweetRequest {
    parent: string;
    tweet: MTweet;
  }
  
  interface MDeleteTweetRequest {
    name: string;
  }
  
  interface MExportTweetsRequest {
    account: string;
    filename: string;
  }
  
  interface MExportTweetsResponse {}
  
  interface MFollowAccountRequest {
    name: string;
  }
  
  interface MFollowAccountResponse {
    followed_accounts: string[];
  }
  
  interface MLikeTweetRequest {
    name: string;
  }
  
  interface MLikeTweetResponse {}
  
  interface MListTweetsRequest {
    parent: string;
    page_size: number;
    page_token: string;
    order_by: string;
  }
  
  interface MListTweetsResponse {
    tweets: MTweet[];
    next_page_token: string;
  }
  
  interface MReTweetRequest {
    name: string;
  }
  
  interface MReTweetResponse {
    tweet: MTweet;
  }
  
  interface MSearchAccountsRequest {
    name: string;
  }
  
  interface MSearchAccountsResponse {
    accounts: string[];
  }
  
  interface MTweet {
    name: string;
    text: string;
    likes: number;
    original_tweet: string;
  }
  
  interface MUpdateAccountRequest {
    account: MAccount;
    update_mask: google.protobuf.FieldMask;
  }
  
  // TwitterForKidsService
  interface STwitterForKidsService {
    CreateAccount(request: MCreateAccountRequest): MAccount;
    UpdateAccount(request: MUpdateAccountRequest): MAccount;
    AddParent(request: MAddParentRequest): MAddParentResponse;
    SearchAccounts(request: MSearchAccountsRequest): MSearchAccountsResponse;
    FollowAccount(request: MFollowAccountRequest): MFollowAccountResponse;
    CreateTweet(request: MCreateTweetRequest): MTweet;
    DeleteTweet(request: MDeleteTweetRequest): google.protobuf.Empty;
    ListTweets(request: MListTweetsRequest): MListTweetsResponse;
    LikeTweet(request: MLikeTweetRequest): MLikeTweetResponse;
    ReTweet(request: MReTweetRequest): MReTweetResponse;
    CreateComment(request: MCreateCommentRequest): MComment;
    ExportTweets(request: MExportTweetsRequest): MExportTweetsResponse;
  }




// Client side
import * as grpc from 'grpc';
import { MCreateTweetRequest, MDeleteTweetRequest, MLikeTweetRequest, MCreateCommentRequest, MTweet } from './child_twitter_pb';
import { STwitterForKidsServiceClient } from './child_twitter_grpc_pb';

// Create a gRPC client to connect to the server
const client = new STwitterForKidsServiceClient('localhost:50051', grpc.credentials.createInsecure());

// Create a post
function createPost(parent: string, postText: string) {
    const request = new MCreatePostRequest();
    request.setParent(parent);
    const post = new MPost();
    post.setText(postText);
    request.setPost(post);

    client.createPost(request, (error, response) => {
        if (!error) {
            console.log(`Post created: ${response.getPost()?.getName()}`);
        } else {
            console.error('Error creating post:', error.message);
        }
    });
}

// Delete a tweet
function deleteTweet(tweetName: string) {
    const request = new MDeleteTweetRequest();
    request.setName(tweetName);

    client.deleteTweet(request, (error) => {
        if (!error) {
            console.log(`Tweet deleted: ${tweetName}`);
        } else {
            console.error('Error deleting tweet:', error.message);
        }
    });
}

// Like a tweet
function likeTweet(tweetName: string) {
    const request = new MLikeTweetRequest();
    request.setName(tweetName);

    client.likeTweet(request, (error) => {
        if (!error) {
            console.log(`Tweet liked: ${tweetName}`);
        } else {
            console.error('Error liking tweet:', error.message);
        }
    });
}

// Reply to a tweet
function replyToTweet(parentTweetName: string, replyText: string) {
    const request = new MCreateCommentRequest();
    request.setParent(parentPostName);
    const comment = new MTweet();
    comment.setText(replyText);
    request.setComment(comment);

    client.createComment(request, (error, response) => {
        if (!error) {
            console.log(`Reply created: ${response.getName()}`);
        } else {
            console.error('Error creating reply:', error.message);
        }
    });
}

// Usage
if (require.main === module) {
    createTweet('parent_account', 'Hello, Twitter for Kids!');
    deleteTweet('tweet123');
    likeTweet('tweet456');
    replyToTweet('tweet789', "That's great!");
}
